include/realtime_tools/async_function_handler.hpp

```diff
@@ -0,0 +1,274 @@
+// Copyright 2024 PAL Robotics S.L.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// \author Sai Kishor Kothakota
+
+#ifndef REALTIME_TOOLS__ASYNC_FUNCTION_HANDLER_HPP_
+#define REALTIME_TOOLS__ASYNC_FUNCTION_HANDLER_HPP_
+
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <limits>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include "rclcpp/duration.hpp"
+#include "rclcpp/logging.hpp"
+#include "rclcpp/time.hpp"
+#include "realtime_tools/thread_priority.hpp"
+
+namespace realtime_tools
+{
+/**
+ * @brief Class to handle asynchronous function calls.
+ * AsyncFunctionHandler is a class that allows the user to have a asynchronous call to the parsed
+ * method and be able to set some thread specific parameters
+ */
+template <typename T>
+class AsyncFunctionHandler
+{
+public:
+  AsyncFunctionHandler() = default;
+
+  ~AsyncFunctionHandler() { stop_async_update(); }
+
+  /// Initialize the AsyncFunctionHandler with the async_function and thread_priority
+  /**
+   * @param async_function Function that will be called asynchronously
+   * If the AsyncFunctionHandler is already initialized and is running, it will throw a runtime
+   * error.
+   * If the parsed functions are not valid, it will throw a runtime error.
+   */
+  void init(
+    std::function<T(const rclcpp::Time &, const rclcpp::Duration &)> async_function,
+    int thread_priority = 50)
+  {
+    if (async_function == nullptr) {
+      throw std::runtime_error(
+        "AsyncFunctionHandler: parsed function to call asynchronously is not valid!");
+    }
+    if (thread_.joinable()) {
+      throw std::runtime_error(
+        "AsyncFunctionHandler: Cannot reinitialize while the thread is "
+        "running. Please stop the async update first!");
+    }
+    async_function_ = async_function;
+    thread_priority_ = thread_priority;
+  }
+
+  /// Initialize the AsyncFunctionHandler with the async_function and trigger_predicate
+  /**
+   * @param async_function Function that will be called asynchronously
+   * @param trigger_predicate Predicate function that will be called to check if the async update
+   * method should be triggered.
```

Comment:
This method also takes a `thread_priority` argument. Let's document that as well?

Reply:
Thanks for pointing out. Done!